### DISCUSSION TITLE
Slow continuous testing on changes

### DISCUSSION THREAD

**Discussion by ennishol** [quarkusio/quarkus#34114] [Q&A] [answered]
Hi,
I have an application where the reload time in continuous test mode is around 20 seconds. I have tried profiling the realoading  but couldn't find any useful information. In the log, it seems that loading my testhelper library is causing the bottleneck, but I'm unsure why it could be happening.

I would appreciate any recommendations or suggestions on how to identify the cause of this issue. It's worth mentioning that it used to work fine with older versions of Quarkus.

Thank you in advance for your assistance.

023-06-17 17:19:35,988 DEBUG [io.qua.boo.cla.QuarkusClassLoader] (Test runner thread) Adding elements io.quarkus.bootstrap.classloading.PathTreeClassPathElement[com.acme:app-library / runtime=true resources=null] to QuarkusClassLoader Deployment Class Loader: TEST
2023-06-17 **17:19:35,989** DEBUG [io.qua.boo.cla.QuarkusClassLoader] (Test runner thread) Adding elements io.quarkus.bootstrap.classloading.PathTreeClassPathElement[com.acme:**testhelper** / runtime=true resources=null] to QuarkusClassLoader Deployment Class Loader: TEST
2023-06-17 **17:19:51,726** DEBUG [io.qua.boo.cla.QuarkusClassLoader] (Test runner thread) Adding elements io.quarkus.bootstrap.classloading.PathTreeClassPathElement[/workspace/app/application/target/test-classes runtime=true resources=null] to QuarkusClassLoader Deployment Class Loader: TEST
2023-06-17 17:19:51,727 DEBUG [io.qua.boo.cla.QuarkusClassLoader] (Test runner thread) Adding elements io.quarkus.bootstrap.classloading.PathTreeClassPathElement[/workspace/app/application/target/classes runtime=true resources=null] to QuarkusClassLoader Deployment Class Loader: TEST

[a comment by an automated account is not shown]

**Comment by stuartwdouglas** [comment]
I think it is actually `/workspace/app/application/target/test-classes` that is slow. Are you running an anti virus by any chance? If so could it be massivly slowing down the scanning of the test classes dir?

**Comment by ennishol** [comment]
@stuartwdouglas  I'm on Ubuntu hyper-v VM and no antivirus in the VM. I did JFR recording and in the slowest part there is a lot of allocation outside of TLABs. Here some screens:
TLAB:
![tlab alloc](https://github.com/quarkusio/quarkus/assets/64908285/79d47930-f8b3-4fb3-bf63-1946cd90b74e)
Memory
![memory profiling](https://github.com/quarkusio/quarkus/assets/64908285/adeddef7-168e-4dcc-b9da-7bfaea9be896)
Methods
![method profiling](https://github.com/quarkusio/quarkus/assets/64908285/d4120288-ff78-4140-89ce-81cf486360a7)

**Comment by ennishol** [comment]
After some debugging I see that a transitive dependency with 4646 classes of testhelper is parsed and transformed in this part of the quarkus code:

![loop](https://github.com/quarkusio/quarkus/assets/64908285/910b968c-efc4-4e59-9f74-b0d296bf1b97)

**Comment by stuartwdouglas** [comment, marked as answer]
Have you explicitly marked this dependency as hot reloadable `quarkus.class-loading.reloadable-artifacts` ?

If not then there is likely a bug here.

**Comment by ennishol** [comment]
@stuartwdouglas Indeed, I have the reloadable setting on. I've completely forgotten about it. It was added because of some class loading issues. But now it seems to work also without it. 
Thank you very much!

**Comment by stuartwdouglas** [comment]
We still should probably do something here to make large test suites work better, but maybe it can wait until someone complains.
